src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java

```diff
@@ -491,4 +496,97 @@ public void testQueryByDynamicProperty() throws Exception {
 
         assertThat((List<String>) doc.read("features[*].id"), contains("SAS1_20180226102021.01"));
     }
+
+    @Test
+    public void testPropertySelectionOnDynamicMerge() throws Exception {
+        DocumentContext result =
+                getAsJSONPath(
+                        "ogc/stac/collections/LANDSAT8/items?fields=properties,-assets,assets.thumbnail.type",
+                        200);
+
+        // empty fields param only mandatory attributes should be there.
+        String id = result.read("features[0].id");
+        String type = result.read("features[0].type");
+        Map<String, Object> geom = result.read("features[0].geometry");
+        JSONArray bbox = result.read("features[0].bbox");
+        Map<String, Object> properties = result.read("features[0].properties");
+        Map<String, Object> assets = result.read("features[0].assets");
+        String thumbnailType = result.read("features[0].assets.thumbnail.type");
+        JSONArray links = result.read("features[0].links");
+
+        assertEquals("LS8_TEST.02", id);
+        assertEquals("Feature", type);
+        assertEquals(2, geom.size());
+        assertEquals(4, bbox.size());
+        assertTrue(properties.size() > 2);
+        assertNotNull(assets);
+        assertEquals(1, assets.size());
+        assertEquals("will replace", thumbnailType);
+        assertNotNull(links);
+        assertFalse(links.isEmpty());
+    }
+
+    @Test
+    public void testPropertySelectionDynamicMerge2() throws Exception {
+        DocumentContext result =
+                getAsJSONPath(
+                        "ogc/stac/collections/LANDSAT8/items?fields=properties,-properties.created,-properties.datetime,-assets.thumbnail",
+                        200);
+
+        // empty fields param only mandatory attributes should be there.
+        String id = result.read("features[0].id");
+        String type = result.read("features[0].type");
+        Map<String, Object> geom = result.read("features[0].geometry");
+        JSONArray bbox = result.read("features[0].bbox");
+        Map<String, Object> properties = result.read("features[0].properties");
+        Map<String, Object> assets = result.read("features[0].assets");
+        JSONArray links = result.read("features[0].links");
+
+        assertEquals("LS8_TEST.02", id);
+        assertEquals("Feature", type);
+        assertEquals(2, geom.size());
+        assertEquals(4, bbox.size());
+        // more then just the two mandatory ones.
+        assertTrue(properties.size() > 2);
```

Comment:
Check some of the actual keys and values are there.

Reply:
fixed